torch/_inductor/fx_passes/group_fusion.py

```diff
@@ -0,0 +1,142 @@
+import operator
+from typing import List
+
+import networkx
+import torch
+
+from ..pattern_matcher import (
+    CallFunctionVarArgs,
+    get_arg_value,
+    config_flag,
+    stable_topological_sort,
+)
+from torch._dynamo.utils import counters
+
+
+def get_nx_graph_from_fx_graph(graph: torch.fx.Graph) -> networkx.DiGraph:
+    G = networkx.DiGraph()
+    for node in graph.nodes:
+        G.add_node(node)
+        for user in node.users:
+            G.add_edge(node, user)
+    return G
+
+
+class GroupFusionPass:
+    def __init__(self, pattern, pair_check, replacement_fn, extra_check=lambda m: True):
+        self.pattern = pattern
+        self.extra_check = extra_check
+        self.pair_check = pair_check
+        self.replacement_fn = replacement_fn
+
+    def apply(self, graph):
+        target_nodes = []
+        for node in graph.nodes:
+            if m := self.pattern.match(node):
+                if self.extra_check(m):
+                    target_nodes.append(node)
+
+        seen_nodes = set()
+
+        for i, target_node in enumerate(target_nodes):
+            nodes_to_fuse = []
+            nx_graph = get_nx_graph_from_fx_graph(graph)
+            if target_node not in seen_nodes:
+                seen_nodes.add(target_node)
+                nodes_to_fuse.append(target_node)
+
+                for j in range(i + 1, len(target_nodes)):
+                    if target_nodes[j] in seen_nodes:
+                        continue
+                    can_be_fused = self.pair_check(target_node, target_nodes[j])
+                    if (
+                        can_be_fused
```

Comment:
We should add a check here that `get_mutation_region_id` is the same for the two nodes.
https://github.com/pytorch/pytorch/blob/e26f5b2ac718995822fc410fbb08ee612ab29efb/torch/_inductor/pattern_matcher.py#L799